### PR TITLE
musescore: support Muse Sounds

### LINKED ISSUE #318210: MuseScore does not detect Muse Sounds Manager
### Describe the bug
`musescore` (`4.3.1`/`4.3.2`/`4.4.0`/`4.4.1`) packaged in `nixpkgs-unstable` does not detect Muse Sounds Manager. The Muse Sounds profile is greyed out and cannot be used; leaving only MuseScore Basic available to use.

### Steps To Reproduce
1. Install `musescore` from `nixpkgs-unstable` (at the time of updating, version `4.3.2`)
2. Use the unofficially packaged [`muse-sounds-manager-nix`](https://github.com/thilobillerbeck/muse-sounds-manager-nix)) flake.
3. Open MuseScore either from the desktop entry or via `mscore` in a terminal, visit `View > Playback Setup` in the top bar, to check for available playback sounds to use.

### Expected behavior
The Muse Sounds profile should be available to use and not greyed out.

### Screenshots
image

### Additional context
I have acknowledged the fact that although the mentioned Muse Sounds Manager package is unofficially packaged, **the package works with the AppImage Version of MuseScore** (packaged officially [here](https://musescore.org/en/download)), so I don't think this is the problem with the Muse Sounds Manager flake.

Below is the screenshot of what the Playback Setup window should look like, taken from the AppImage version while opening the same score as the previous screenshot.

image

I have tried reinstalling Muse Sounds Manager itself and reinstalling each Muse Sounds, but the issue persists.

My entire system configuration files are [here](https://github.com/fdnt7/nix-config).

### Notify maintainers

@vandenoever @doronbehar 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.9.3, NixOS, 24.11 (Vicuña), 24.11.20240605.e8057b6`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.2`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/store/8s55w0927lh3mdbkxf434zb0c5hqsz8z-source`
```
---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by orivej**
## Description of changes

After this change Muse Sounds are always active in MuseScore, even if it is launched from PATH or via File -> New

Fixes #318210

/cc @doronbehar

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/applications/audio/musescore/default.nix
Comment (doronbehar) [inline pkgs/applications/audio/musescore/default.nix:157]: Why is it an issue to wrapQtApps on non linux systems? Again please.

File: pkgs/applications/audio/musescore/default.nix
Comment (doronbehar) [inline pkgs/applications/audio/musescore/default.nix:161]: It feels a bit odd to do that. If I understand correctly, using `makeQtWrapper` with `--argv mscore` should be equivalent?

Comment (doronbehar) [reply pkgs/applications/audio/musescore/default.nix:161]: > It feels a bit odd to do that. If I understand correctly, using `makeQtWrapper` with `--argv mscore` should be equivalent?

Oh wait never mind, I understand now - we can't wrap `.mscore-wrapped`, that's the main point.

Comment (doronbehar) [reply pkgs/applications/audio/musescore/default.nix:161]: I'd still like to put that wrapped `mscore` somewhere else - how about `$out/libexec/mscore`?

File: pkgs/applications/audio/musescore/default.nix
Comment (orivej) [reply pkgs/applications/audio/musescore/default.nix:157]: After this PR:
1) On Linux, we disable the default `wrapQtAppsHook` and wrap the binary explicitly.
2) On macOS, nothing changes.

If I did this unconditionally:
```
  dontWrapQtApps = true;
  postFixup = ''
    mv $out/bin/mscore $out
    makeQtWrapper $out/mscore $out/bin/mscore
  '';
```
I might break MuseScore on macOS because there `wrapQtAppsHook` probably wraps not only `$out/bin/mscore` but also something in `$out/Applications/`, and users may need that.

This PR needs testing on macOS and possibly some changes to be enabled for macOS.

File: pkgs/applications/audio/musescore/default.nix
Comment (orivej) [reply pkgs/applications/audio/musescore/default.nix:161]: `makeQtWrapper --argv0 $out/mscore` (we need argv0 to end with "/mscore") would be equivalent, but then `--argv0` is unnecessary because this is how executing programs by absolute path normally works: their argv0 is that absolute path.

Moving `$out/bin/mscore` out of `$out/bin` is necessary because both the end-user wrapper and the wrapped binary must be named `mscore`, and the wrapper should be left in `$out/bin/mscore`.

File: pkgs/applications/audio/musescore/default.nix
Comment (orivej) [reply pkgs/applications/audio/musescore/default.nix:157]: macOS needs wrapQtAppsHook for the same reason Linux does: Qt has to find its plugins with the help of QT_PLUGIN_PATH and NIXPKGS_QT6_QML_IMPORT_PATH environment variables.

Comment (doronbehar) [reply pkgs/applications/audio/musescore/default.nix:157]: > macOS needs wrapQtAppsHook for the same reason Linux does: Qt has to find its plugins with the help of QT_PLUGIN_PATH and NIXPKGS_QT6_QML_IMPORT_PATH environment variables.

That's what I thought, otherwise `wrapQtAppsHook` would have been set to `null` on Darwin platforms.

File: pkgs/applications/audio/musescore/default.nix
Comment (orivej) [reply pkgs/applications/audio/musescore/default.nix:161]: I can move it to libexec.

Comment (doronbehar) [reply pkgs/applications/audio/musescore/default.nix:161]: > Moving `$out/bin/mscore` out of `$out/bin` is necessary because both the end-user wrapper and the wrapped binary must be named `mscore`, and the wrapper should be left in `$out/bin/mscore`.

I understand that, but you don't state that there is a problem with using `$out/libexec/mscore` as the path to the wrapped executable right?

Comment (doronbehar) [reply pkgs/applications/audio/musescore/default.nix:161]: > I can move it to libexec.

Wait maybe don't do that, due to this:

https://github.com/NixOS/nixpkgs/blob/31437c831582f3b7c54db408fbcbd3e9ae264648/pkgs/development/libraries/qt-5/hooks/wrap-qt-apps-hook.sh#L79

Let's go for `$out/lib/mscore`

File: pkgs/applications/audio/musescore/default.nix
Comment (doronbehar) [inline pkgs/applications/audio/musescore/default.nix:158]: Didn't we agree that this should be done for Darwin as well?

```suggestion
  dontWrapQtApps = true;
  postFixup = ''
```

We need to check somehow though if a similar action needs to be done in `$out/Applications`. I have access to the Darwin build box at `darwin-build-box.nix-community.org` but apparently musescore is broken on Darwin since the last build:

https://cache.nixos.org/log/28xjqwnby06ihpjqc0yq8alnqprai8p2-musescore-4.4.1.drv

https://hydra.nixos.org/build/272681576

File: pkgs/applications/audio/musescore/default.nix
Comment (orivej) [reply pkgs/applications/audio/musescore/default.nix]: No, I did not think of it myself but libexec is exactly the right place, thank you for the suggestion!
wrap-qt-apps-hook touches it exactly because it is a place for executables, but we are not concerned with that because we use dontWrapQtApps.

Comment (doronbehar) [reply pkgs/applications/audio/musescore/default.nix]: > No, I did not think of it myself but libexec is exactly the right place, thank you for the suggestion! wrap-qt-apps-hook touches it exactly because it is a place for executables, but we are not concerned with that because we use dontWrapQtApps.

Right, I was just thinking about it :).

File: pkgs/applications/audio/musescore/default.nix
Comment (orivej) [reply pkgs/applications/audio/musescore/default.nix:158]: We will also need someone to test how MuseScore behaves on macOS with respect to Muse Sounds. For example it is possible that libMuseSamplerCoreLib on macOS does not check the executable name, then Muse Sounds should already work there and this PR will not need further adjustments. I'd like to keep this PR Linux-only.

Comment (doronbehar) [reply pkgs/applications/audio/musescore/default.nix:158]: > We will also need someone to test how MuseScore behaves on macOS with respect to Muse Sounds. For example it is possible that libMuseSamplerCoreLib on macOS does not check the executable name, then Muse Sounds should already work there and this PR will not need further adjustments. I'd like to keep this PR Linux-only.

OK That's an option, but let's write a TODO comment for that.

Comment (doronbehar) [reply pkgs/applications/audio/musescore/default.nix:158]: In the meantime the 4.4.2 build of musescore on Darwin seems to go on well (it was probably an upstream regression in 4.4.1).

**Comment by fdnt7** [comment]
Hey, I've built and run the package (on NixOS x86_64-linux). Muse Sounds is enabled, both on the first instance and the subsequent instances opened via `File > New` / `File > Open Recent`. I haven't tested `File > Open` because #342089 not being merged here. Also none of the panels that dock appear when I try to toggle it (like Timeline, Mixer, etc) but this seems unrelated also since it also happens with nixpkgs's 4.4.1.

**Comment by doronbehar** [comment]
> Hey, I've built and run the package (on NixOS x86_64-linux). Muse Sounds is enabled, both on the first instance and the subsequent instances opened via `File > New` / `File > Open Recent`. I haven't tested `File > Open` because #342089 not being merged here. Also none of the panels that dock appear when I try to toggle it (like Timeline, Mixer, etc) but this seems unrelated also since it also happens with nixpkgs's 4.4.1.

Thanks for your test. That's a sufficient approval. PR link fix: you meant https://github.com/NixOS/nixpkgs/pull/342316 .

**Comment by doronbehar** [comment]
Also the Darwin build of 4.4.2 in Nixpkgs' current `master` branch works, so I expect no regressions in that from any of the musescore PRs open currently - as they both aim Linux only.

**Comment by orivej** [comment]
> Also none of the panels that dock appear when I try to toggle it (like Timeline, Mixer, etc) but this seems unrelated also since it also happens with nixpkgs's 4.4.1.

This affects me too. Seems like incompatibility with our version of Qt. I hope to fix this as well, but I've made no clear progress so far

File: pkgs/applications/audio/musescore/default.nix
Comment (orivej) [reply pkgs/applications/audio/musescore/default.nix:160]: Added TODO

**Review by doronbehar — APPROVED** [review]
Should be good to go once CI is green.

**Comment by doronbehar** [comment]
BTW @orivej would you like to join the maintainers list of `musescore`? Please don't do that in a separate PR haha I will lose my patience if I had to wait for another ofborg evaluation.

**Comment by doronbehar** [comment]
Thanks again @orivej for all of those patches. Very appreciated :heart:.

**Comment by orivej** [comment]
Thank you for detailed feedback!

**Comment by adam248** [comment]
> > Also none of the panels that dock appear when I try to toggle it (like Timeline, Mixer, etc) but this seems unrelated also since it also happens with nixpkgs's 4.4.1.
> 
> This affects me too. Seems like incompatibility with our version of Qt. I hope to fix this as well, but I've made no clear progress so far

I am also affected by this issue and there is no specific error when trying to toggle the mixer.
4.4.1 and 4.4.2 tested.
Do we have a separate issue on git hub  to track this bug? (i tried searching but only found https://github.com/NixOS/nixpkgs/issues/180841)
@orivej @fdnt7 ?

**Pull request merged**
